static/app/utils/performance/contexts/genericQueryBatcher.tsx

```diff
@@ -0,0 +1,240 @@
+import {createContext, Fragment, Ref, useEffect, useRef} from 'react';
+
+import {Client} from 'app/api';
+import useApi from 'app/utils/useApi';
+
+import {createDefinedContext} from './utils';
+
+type QueryObject = {
+  query: {
+    [k: string]: any;
+  };
+}; // TODO(k-fish): Fix to ensure exact types for all requests. Simplified type for now, need to pull this in from events file.
+
+type BatchQueryDefinition = {
+  // Intermediate promise functions
+  resolve: (value: any) => void;
+  reject: (reason?: string) => void;
+
+  batchProperty: string;
+  requestQueryObject: QueryObject;
+  path: string;
+  api: Client;
+};
+
+type QueryBatch = {
+  addQuery: (q: BatchQueryDefinition, id: symbol) => void;
+};
+
+const [GenericQueryBatcherProvider, _useGenericQueryBatcher] =
+  createDefinedContext<QueryBatch>({
+    name: 'GenericQueryBatcherContext',
+  });
+
+function mergeKey(query: BatchQueryDefinition) {
+  return `${query.batchProperty}.${query.path}`;
+}
+
+type MergeMap = Record<string, BatchQueryDefinition[]>;
+
+// Builds a map that will contain an array of query definitions by mergeable key (using batch property and path)
+function queriesToMap(collectedQueries: Record<symbol, BatchQueryDefinition>) {
+  const keys = Reflect.ownKeys(collectedQueries);
+  if (!keys.length) {
+    return false;
+  }
+  const mergeMap: MergeMap = {};
+
+  keys.forEach(async key => {
+    const query = collectedQueries[key];
+    mergeMap[mergeKey(query)] = mergeMap[mergeKey(query)] || [];
+    mergeMap[mergeKey(query)].push(query);
+    delete collectedQueries[key];
+  });
+
+  return mergeMap;
+}
+
+function requestFunction(api: Client, path: string, queryObject: QueryObject) {
+  return api.requestPromise(path, queryObject);
+}
+
+function _handleUnmergeableQuery(queryDefinition: BatchQueryDefinition) {
+  const result = requestFunction(
+    queryDefinition.api,
+    queryDefinition.path,
+    queryDefinition.requestQueryObject
+  );
+  queryDefinition.resolve(result);
+}
+
+function _handleUnmergeableQueries(mergeMap: MergeMap) {
+  Object.keys(mergeMap).forEach(async k => {
+    // Using async forEach to ensure calls start in parallel.
+    const mergeList = mergeMap[k];
+
+    if (mergeList.length !== 1) {
+      return;
+    }
+
+    const [queryDefinition] = mergeList;
+    _handleUnmergeableQuery(queryDefinition);
+  });
+}
+
+function _handleMergeableQueries(mergeMap: MergeMap) {
+  Object.keys(mergeMap).forEach(async k => {
+    const mergeList = mergeMap[k];
+
+    if (mergeList.length <= 1) {
+      return;
+    }
+
+    const [exampleDefinition] = mergeList;
+    const batchProperty = exampleDefinition.batchProperty;
+    const requestQueryObject = {...exampleDefinition.requestQueryObject};
+
+    const batchValues = mergeList.map(q => {
+      const batchFieldValue = q.requestQueryObject.query[batchProperty];
+      if (Array.isArray(batchFieldValue)) {
+        return batchFieldValue[0];
+      }
```

Comment:
No, since then it picks the first. This mostly just works around the type (when I was type checking) since it can be `string` or `string[]` with one value. We could also check multi values and have it fallback to the regular query for those ones.